Find 3 users who are closest to you in terms of age, gender, and occupation.

For each of those 3 users:

A: What are their top 3 (favorite) films?
B: What are their bottom 3 (least favorite) films?
Based on the movie values in those 6 tables (3 users X (favorite + least favorite)), choose a user that you feel is most like you. Feel free to note any outliers (e.g., "I mostly identify with user 123, except I did not like "Ghost" at all"). You can investigate more than just the top 3 and bottom 3 movies to find your best match.

This user is the substitute you.

In [1]:
user1 = open("/content/drive/MyDrive/DATA440Fall2023/u.user.txt","r")

In [2]:
user1.readlines(20)

['1|24|M|technician|85711\n']

In [3]:

user1 = open("/content/drive/MyDrive/DATA440Fall2023/u.user.txt","r")
while True:
  content=user1.readline()
  if not content:
    break

  print(content)
user1.close()

1|24|M|technician|85711

2|53|F|other|94043

3|23|M|writer|32067

4|24|M|technician|43537

5|33|F|other|15213

6|42|M|executive|98101

7|57|M|administrator|91344

8|36|M|administrator|05201

9|29|M|student|01002

10|53|M|lawyer|90703

11|39|F|other|30329

12|28|F|other|06405

13|47|M|educator|29206

14|45|M|scientist|55106

15|49|F|educator|97301

16|21|M|entertainment|10309

17|30|M|programmer|06355

18|35|F|other|37212

19|40|M|librarian|02138

20|42|F|homemaker|95660

21|26|M|writer|30068

22|25|M|writer|40206

23|30|F|artist|48197

24|21|F|artist|94533

25|39|M|engineer|55107

26|49|M|engineer|21044

27|40|F|librarian|30030

28|32|M|writer|55369

29|41|M|programmer|94043

30|7|M|student|55436

31|24|M|artist|10003

32|28|F|student|78741

33|23|M|student|27510

34|38|F|administrator|42141

35|20|F|homemaker|42459

36|19|F|student|93117

37|23|M|student|55105

38|28|F|other|54467

39|41|M|entertainment|01040

40|38|M|scientist|27514

41|33|M|engineer|80525

42|30|M|administrator|1787

In [4]:


user1 = open("/content/drive/MyDrive/DATA440Fall2023/u.user.txt","r")
while True:
  content=user1.readline()
  if '21' in content and 'F' in content and 'student' in content:
    print(content)
  #print(type(content)) #content is a string
  if not content:
    break

user1.close()

198|21|F|student|55414

228|21|F|student|22003

324|21|F|student|02176

477|23|F|student|02125

496|21|F|student|55414

618|15|F|student|44212

642|18|F|student|95521

705|21|F|student|54494

813|14|F|student|02136

921|20|F|student|98801



In [ ]:
'''
324|21|F|student|02176

477|23|F|student|02125

496|21|F|student|55414
'''

#lets pick these


In [5]:
user1 = open("/content/drive/MyDrive/DATA440Fall2023/u.user.txt","r")
data1 = open("/content/drive/MyDrive/DATA440Fall2023/u.data.txt","r")
item1 = open("/content/drive/MyDrive/DATA440Fall2023/u.item.txt","r")



In [6]:
user1 = open("/content/drive/MyDrive/DATA440Fall2023/u.user.txt","r")
while True:
  content=user1.readline()
  if '21' in content and 'F' in content and 'student' in content:
    print(content)
  #print(type(content)) #content is a string
  if not content:
    break

user1.close()

198|21|F|student|55414

228|21|F|student|22003

324|21|F|student|02176

477|23|F|student|02125

496|21|F|student|55414

618|15|F|student|44212

642|18|F|student|95521

705|21|F|student|54494

813|14|F|student|02136

921|20|F|student|98801



In [ ]:
#user 324:
#top 3:
# 324	748	5	880575108
# 324	471	5	880575412
# 324	298	5	880575493

# bottom 3:
# 324	877	1	880575163
# 324	283	3	880575531
# 324	332	3	880574766

In [ ]:
#user 477:
#top 3:

477	815	5	875941763

477	709	5	875941763

477	88	5	875941085

# bottom 3:

# 477	280	4	875941022

# 477	781	4	875941191

# 477	15	4	875941863


# ---

#user 496:


#top 3:


496	727	5	876072633 - 727|Immortal Beloved (1994)|01-Jan-1994||http://us.imdb.com/M/title-exact?Immortal%20Beloved%20(1994)|0|0|0|0|0|0|0|0|1|0|0|0|0|0|1|0|0|0|0


496	921	5	876072633 - 921|Farewell My Concubine (1993)|01-Jan-1993||http://us.imdb.com/M/title-exact?Ba%20Wang%20Bie%20Ji%20(1993)|0|0|0|0|0|0|0|0|1|0|0|0|0|0|1|0|0|0|0

496	42	5	876066676 - 42|Clerks (1994)|01-Jan-1994||http://us.imdb.com/M/title-exact?Clerks%20(1994)|0|0|0|0|0|1|0|0|0|0|0|0|0|0|0|0|0|0|0


# bottom 3:

496	97	1	876066848 - 97|Dances with Wolves (1990)|01-Jan-1990||http://us.imdb.com/M/title-exact?Dances%20with%20Wolves%20(1990)|0|0|1|0|0|0|0|0|1|0|0|0|0|0|0|0|0|0|1

496	1444	1	876066465 - 1444|That Darn Cat! (1965)|01-Jan-1965||http://us.imdb.com/Title?That+Darn+Cat%21+(1965)|0|0|0|0|1|1|0|0|0|0|0|0|0|1|0|0|0|0|0

496	136	1	876066424 - 136|Mr. Smith Goes to Washington (1939)|01-Jan-1939||http://us.imdb.com/M/title-exact?Mr.%20Smith%20Goes%20to%20Washington%20(1939)|0|0|0|0|0|0|0|0|1|0|0|0|0|0|0|0|0|0|0


In [ ]:
# five users that love 921, just like my subsitute user
94	921	5	891725332
295	727	5	879517682

In [7]:
data1 = open("/content/drive/MyDrive/DATA440Fall2023/u.data.txt","r")
while True:
  content2 = data1.readline()
  content2_list = content2.split('\t')
  if '496' in content2_list[0] and '5' in content2_list[2]:
    print(content2)
  if not content2:
    break
#data1.close()

496	727	5	876072633

496	921	5	876072633

496	42	5	876066676

496	87	5	876073616

496	39	5	876072633

496	181	5	876064168

496	774	5	876066424

496	133	5	876066567

496	532	5	876072633

496	561	5	876068582

496	89	5	876072633

496	50	5	876072633

496	173	5	876065349

496	172	5	876065558

496	191	5	876072632

496	56	5	876066009

496	10	5	876064845

496	652	5	876065693

496	277	5	876072633

496	559	5	876068153

496	721	5	876067215

496	190	5	876072632

496	217	5	876073320



In [8]:
# all the ratings for the films that my subsitute has not seen
data1 = open("/content/drive/MyDrive/DATA440Fall2023/u.data.txt","r")
while True:
  content2 = data1.readline()
  content2_list = content2.split('\t')
  if '496' not in content2_list[0]:
    print(content2)
  if not content2:
    break
#data1.close()

Streaming output truncated to the last 5000 lines.

731	520	4	886186567

892	797	4	886610372

876	286	5	879428072

758	315	5	883793836

606	1149	4	880925289

881	1217	5	876538506

271	429	4	885848672

876	523	5	879428378

924	402	3	886759965

916	66	3	880845264

770	257	4	875972059

927	1284	4	879181133

916	421	5	880844291

745	603	4	880123243

586	939	4	884064459

940	7	4	885921597

815	239	5	878694563

892	175	4	886608559

921	69	4	879380862

919	678	2	875288253

848	132	5	887038197

546	300	3	885139842

919	95	4	875921182

321	59	4	879440687

597	748	5	875339041

537	69	2	886031178

785	137	2	879438810

389	618	4	880088115

484	38	4	891195532

933	226	2	874854874

746	83	4	885075497

574	311	4	891279410

637	127	2	882901356

856	328	3	891489478

862	202	5	879304624

790	210	4	885155209

618	776	2	891307098

133	271	5	890588766

495	393	5	888635339

871	305	3	888192475

728	748	3	879442532

682	8	3	888521833

469	487	5	879524178

883	241	4	891696714

721	1065	5	877147383

201	28	3	8

In [9]:
data1 = open("/content/drive/MyDrive/DATA440Fall2023/u.data.txt","r")

content2 = data1.readline()
content2_list = content2.split('\t')
print(len(content2_list))
  # if '496' in content2_list[0] and '5' in content2_list[2]:
  #   print(content2)
  # if not content2:
  #   break
#data1.close()

4


favorite film: Farewell My Concubine

least favorite film: 99|Snow White and the Seven Dwarfs (1937)|01-Jan-1937||http://us.imdb.com/M/title-exact?Snow%20White%20and%20the%20Seven%20Dwarfs%20(1937)|0|0|0|1|1|0|0|0|0|0|0|0|1|0|0|0|0|0|0


In [27]:
item1 = open("/content/drive/MyDrive/DATA440Fall2023/u.item.txt","r",encoding = "ISO-8859-1")
while True:
  content=item1.readline()
  if 'Snow White and the Seven Dwarfs' in content:
    print(content)
  if not content:
    break

99|Snow White and the Seven Dwarfs (1937)|01-Jan-1937||http://us.imdb.com/M/title-exact?Snow%20White%20and%20the%20Seven%20Dwarfs%20(1937)|0|0|0|1|1|0|0|0|0|0|0|0|1|0|0|0|0|0|0



In [ ]:
item1 = open("/content/drive/MyDrive/DATA440Fall2023/u.item.txt","r")
for line in item1:
  print(line)
  #user_id, movie_id, rating, _ = map(int, line.strip().split('\t'))

In [ ]:
data1 = open("/content/drive/MyDrive/DATA440Fall2023/u.data.txt","r")
while True:
  content2 = data1.readline()
  content2_list = content2.split('\t')
  for i in content2:

  if '727' in content2_list[1] and '5' in content2_list[2]:
    print(content2)
  if not content2:
    break
#data1.close()

In [12]:
data1 = open("/content/drive/MyDrive/DATA440Fall2023/u.data.txt","r")
for line in data1:
  user_id, movie_id, rating, _ = map(int, line.strip().split('\t'))

user_id, movie_id, rating

(12, 203, 3)

In [13]:
nested_dict = {}

with open("/content/drive/MyDrive/DATA440Fall2023/u.data.txt","r") as file:
    for line in file:
        user_id, movie_id, rating, _ = map(int, line.strip().split('\t'))
        nested_dict.setdefault(user_id, {})[movie_id] = rating
    nested_dict

In [14]:
critics = nested_dict

In [18]:
def getRecommendations(prefs, person, similarity=sim_pearson):
    '''
    Gets recommendations for a person by using a weighted average
    of every other user's rankings
    '''

    totals = {}
    simSums = {}
    for other in prefs:
    # Don't compare me to myself
        if other == person:
            continue
        sim = similarity(prefs, person, other)
        # Ignore scores of zero or lower
        if sim <= 0:
            continue
        for item in prefs[other]:
            # Only score movies I haven't seen yet
            if item not in prefs[person] or prefs[person][item] == 0:
                # Similarity * Score
                totals.setdefault(item, 0)
                # The final score is calculated by multiplying each item by the
                #   similarity and adding these products together
                totals[item] += prefs[other][item] * sim
                # Sum of similarities
                simSums.setdefault(item, 0)
                simSums[item] += sim
    # Create the normalized list
    rankings = [(total / simSums[item], item) for (item, total) in
                totals.items()]
    # Return the sorted list
    rankings.sort()
    rankings.reverse()
    return rankings

In [24]:
# list of the bottom 5 recommendations
getRecommendations(critics, 496)[0:5]

[(5.000000000000001, 1500),
 (5.000000000000001, 850),
 (5.0, 1656),
 (5.0, 1617),
 (5.0, 1612)]

In [17]:
def sim_pearson(prefs, p1, p2):
    '''
    Returns the Pearson correlation coefficient for p1 and p2.
    '''

    # Get the list of mutually rated items
    si = {}
    for item in prefs[p1]:
        if item in prefs[p2]:
            si[item] = 1

    # If they are no ratings in common, return 0
    if len(si) == 0:
        return 0

    # Sum calculations
    n = len(si)

    # Sums of all the preferences
    sum1 = sum([prefs[p1][it] for it in si])
    sum2 = sum([prefs[p2][it] for it in si])

    # Sums of the squares
    sum1Sq = sum([pow(prefs[p1][it], 2) for it in si])
    sum2Sq = sum([pow(prefs[p2][it], 2) for it in si])

    # Sum of the products
    pSum = sum([prefs[p1][it] * prefs[p2][it] for it in si])

    # Calculate r (Pearson score)
    num = pSum - sum1 * sum2 / n
    den = sqrt((sum1Sq - pow(sum1, 2) / n) * (sum2Sq - pow(sum2, 2) / n))
    if den == 0:
        return 0
    r = num / den
    return r

In [16]:
from math import sqrt

In [19]:
def topMatches(
    prefs,
    person,
    n=5,
    similarity=sim_pearson,
):
    '''
    Returns the best matches for person from the prefs dictionary.
    Number of results and similarity function are optional params.
    '''

    scores = [(similarity(prefs, person, other), other) for other in prefs
              if other != person]
    scores.sort()
    scores.reverse()
    return scores[0:n]

In [20]:
topMatches(critics, 496, n=5)

[(1.000000000000004, 644),
 (1.000000000000004, 564),
 (1.0000000000000004, 791),
 (1.0, 939),
 (1.0, 842)]

In [ ]:
#user 477:
#top 3:

# 477	815	5	875941763 - 815|One Fine Day (1996)|30-Nov-1996||http://us.imdb.com/M/title-exact?One%20Fine%20Day%20(1996)|0|0|0|0|0|0|0|0|1|0|0|0|0|0|1|0|0|0|0


# 477	709	5	875941763 - 709|Strictly Ballroom (1992)|01-Jan-1992||http://us.imdb.com/M/title-exact?Strictly%20Ballroom%20(1992)|0|0|0|0|0|1|0|0|0|0|0|0|0|0|1|0|0|0|0


# 477	88	5	875941085 - 88|Sleepless in Seattle (1993)|01-Jan-1993||http://us.imdb.com/M/title-exact?Sleepless%20in%20Seattle%20(1993)|0|0|0|0|0|1|0|0|0|0|0|0|0|0|1|0|0|0|0


# bottom 3:

# 477	280	4	875941022 - 280|Up Close and Personal (1996)|01-Mar-1996||http://us.imdb.com/M/title-exact?Up%20Close%20and%20Personal%20(1996)|0|0|0|0|0|0|0|0|1|0|0|0|0|0|1|0|0|0|0


# 477	781	4	875941191 - 781|French Kiss (1995)|01-Jan-1995||http://us.imdb.com/M/title-exact?French%20Kiss%20(1995)|0|0|0|0|0|1|0|0|0|0|0|0|0|0|1|0|0|0|0


# 477	15	4	875941863 - 15|Mr. Holland's Opus (1995)|29-Jan-1996||http://us.imdb.com/M/title-exact?Mr.%20Holland's%20Opus%20(1995)|0|0|0|0|0|0|0|0|1|0|0|0|0|0|0|0|0|0|0


In [ ]:
#user 324:
#top 3:
# 324	748	5	880575108 - 748|Saint, The (1997)|14-Mar-1997||http://us.imdb.com/M/title-exact?Saint%2C%20The%20(1997)|0|1|0|0|0|0|0|0|0|0|0|0|0|0|1|0|1|0|0

# 324	471	5	880575412 - 471|Courage Under Fire (1996)|08-Mar-1996||http://us.imdb.com/M/title-exact?Courage%20Under%20Fire%20(1996)|0|0|0|0|0|0|0|0|1|0|0|0|0|0|0|0|0|1|0

# 324	298	5	880575493 - 298|Face/Off (1997)|27-Jun-1997||http://us.imdb.com/M/title-exact?Face/Off+(1997)|0|1|0|0|0|0|0|0|0|0|0|0|0|0|0|1|1|0|0


# bottom 3:
# 324	877	1	880575163 - 877|Excess Baggage (1997)|01-Jan-1997||http://us.imdb.com/M/title-exact?Excess+Baggage+(1997)|0|0|1|0|0|0|0|0|0|0|0|0|0|0|1|0|0|0|0

# 324	283	3	880575531 - 283|Emma (1996)|02-Aug-1996||http://us.imdb.com/M/title-exact?Emma%20(1996)|0|0|0|0|0|0|0|0|1|0|0|0|0|0|1|0|0|0|0

# 324	332	3	880574766 - 332|Kiss the Girls (1997)|01-Jan-1997||http://us.imdb.com/M/title-exact?Kiss+the+Girls+(1997)|0|0|0|0|0|0|1|0|1|0|0|0|0|0|0|0|1|0|0


In [55]:
item1 = open("/content/drive/MyDrive/DATA440Fall2023/u.item.txt","r", encoding = "ISO-8859-1")
for i in item1:
  content3 = item1.readline()
content3
  #print(type(content3))
content3_list = content3.split('\t')
#print(content3_list)

In [56]:
content3

'1682|Scream of Stone (Schrei aus Stein) (1991)|08-Mar-1996||http://us.imdb.com/M/title-exact?Schrei%20aus%20Stein%20(1991)|0|0|0|0|0|0|0|0|1|0|0|0|0|0|0|0|0|0|0\n'

In [40]:
item1 = open("/content/drive/MyDrive/DATA440Fall2023/u.item.txt","r", encoding = "ISO-8859-1")
while True:
  content3 = item1.readline()
  #print(type(content3))
  content3_list = content3.split('\t')
  #print(content3_list)
  if 'Farewell My Concubine' in content3_list[0]:
    print(content3)
  if not content3:
    break
item1.close()


921|Farewell My Concubine (1993)|01-Jan-1993||http://us.imdb.com/M/title-exact?Ba%20Wang%20Bie%20Ji%20(1993)|0|0|0|0|0|0|0|0|1|0|0|0|0|0|1|0|0|0|0



In [ ]:
#user 496 is my subsitute, we both gave Farewell My Concubine (1993) a five rating


#even though that was the only movie we had in common

In [ ]:
#user 496:


#top 3:


# 496	727	5	876072633 - 727|Immortal Beloved (1994)|01-Jan-1994||http://us.imdb.com/M/title-exact?Immortal%20Beloved%20(1994)|0|0|0|0|0|0|0|0|1|0|0|0|0|0|1|0|0|0|0


# 496	921	5	876072633 - 921|Farewell My Concubine (1993)|01-Jan-1993||http://us.imdb.com/M/title-exact?Ba%20Wang%20Bie%20Ji%20(1993)|0|0|0|0|0|0|0|0|1|0|0|0|0|0|1|0|0|0|0

# 496	42	5	876066676 - 42|Clerks (1994)|01-Jan-1994||http://us.imdb.com/M/title-exact?Clerks%20(1994)|0|0|0|0|0|1|0|0|0|0|0|0|0|0|0|0|0|0|0


# bottom 3:

# 496	97	1	876066848 - 97|Dances with Wolves (1990)|01-Jan-1990||http://us.imdb.com/M/title-exact?Dances%20with%20Wolves%20(1990)|0|0|1|0|0|0|0|0|1|0|0|0|0|0|0|0|0|0|1

# 496	1444	1	876066465 - 1444|That Darn Cat! (1965)|01-Jan-1965||http://us.imdb.com/Title?That+Darn+Cat%21+(1965)|0|0|0|0|1|1|0|0|0|0|0|0|0|1|0|0|0|0|0

# 496	136	1	876066424 - 136|Mr. Smith Goes to Washington (1939)|01-Jan-1939||http://us.imdb.com/M/title-exact?Mr.%20Smith%20Goes%20to%20Washington%20(1939)|0|0|0|0|0|0|0|0|1|0|0|0|0|0|0|0|0|0|0

A: Which 5 users are most correlated to the substitute you (i.e., which 5 users rate movies most similarly to the substitute you?)


B: Which 5 users are least correlated (i.e., negative correlation)?

In [ ]:
def sim_pearson(prefs, p1, p2):
    '''
    Returns the Pearson correlation coefficient for p1 and p2.
    '''

    # Get the list of mutually rated items
    si = {}
    for item in prefs[p1]:
        if item in prefs[p2]:
            si[item] = 1

    # If they are no ratings in common, return 0
    if len(si) == 0:
        return 0

    # Sum calculations
    n = len(si)

    # Sums of all the preferences
    sum1 = sum([prefs[p1][it] for it in si])
    sum2 = sum([prefs[p2][it] for it in si])

    # Sums of the squares
    sum1Sq = sum([pow(prefs[p1][it], 2) for it in si])
    sum2Sq = sum([pow(prefs[p2][it], 2) for it in si])

    # Sum of the products
    pSum = sum([prefs[p1][it] * prefs[p2][it] for it in si])

    # Calculate r (Pearson score)
    num = pSum - sum1 * sum2 / n
    den = sqrt((sum1Sq - pow(sum1, 2) / n) * (sum2Sq - pow(sum2, 2) / n))
    if den == 0:
        return 0
    r = num / den
    return r

def topMatches(
    prefs,
    person,
    n=5,
    similarity=sim_pearson,
):
    '''
    Returns the best matches for person from the prefs dictionary.
    Number of results and similarity function are optional params.
    '''

    scores = [(similarity(prefs, person, other), other) for other in prefs
              if other != person]
    scores.sort()
    scores.reverse()
    return scores[0:n]

Provide a list of the top 5 recommendations for films that the substitute you should see.

Provide a list of the bottom 5 recommendations (i.e., films the substitute you is almost certain to hate).

In [ ]:
def sim_pearson(prefs, p1, p2):
    '''
    Returns the Pearson correlation coefficient for p1 and p2.
    '''

    # Get the list of mutually rated items
    si = {}
    for item in prefs[p1]:
        if item in prefs[p2]:
            si[item] = 1

    # If they are no ratings in common, return 0
    if len(si) == 0:
        return 0

    # Sum calculations
    n = len(si)

    # Sums of all the preferences
    sum1 = sum([prefs[p1][it] for it in si])
    sum2 = sum([prefs[p2][it] for it in si])

    # Sums of the squares
    sum1Sq = sum([pow(prefs[p1][it], 2) for it in si])
    sum2Sq = sum([pow(prefs[p2][it], 2) for it in si])

    # Sum of the products
    pSum = sum([prefs[p1][it] * prefs[p2][it] for it in si])

    # Calculate r (Pearson score)
    num = pSum - sum1 * sum2 / n
    den = sqrt((sum1Sq - pow(sum1, 2) / n) * (sum2Sq - pow(sum2, 2) / n))
    if den == 0:
        return 0
    r = num / den
    return r

def topMatches(
    prefs,
    person,
    n=5,
    similarity=sim_pearson,
):
    '''
    Returns the best matches for person from the prefs dictionary.
    Number of results and similarity function are optional params.
    '''

    scores = [(similarity(prefs, person, other), other) for other in prefs
              if other != person]
    scores.sort()
    scores.reverse()
    return scores[0:n]

In [ ]:
from math import sqrt

def sim_distance(prefs, p1, p2):
    '''
    Returns a distance-based similarity score for person1 and person2.
    '''

    # Get the list of shared_items
    si = {}
    for item in prefs[p1]:
        if item in prefs[p2]:
            si[item] = 1

    # If they have no ratings in common, return 0
    if len(si) == 0:
        return 0

    # Add up the squares of all the differences
    sum_of_squares = sum([pow(prefs[p1][item] - prefs[p2][item], 2)
                          for item in si])

    return 1 / (1 + sqrt(sum_of_squares))

def sim_pearson(prefs, p1, p2):
    '''
    Returns the Pearson correlation coefficient for p1 and p2.
    '''

    # Get the list of mutually rated items
    si = {}
    for item in prefs[p1]:
        if item in prefs[p2]:
            si[item] = 1

    # If they are no ratings in common, return 0
    if len(si) == 0:
        return 0

    # Sum calculations
    n = len(si)

    # Sums of all the preferences
    sum1 = sum([prefs[p1][it] for it in si])
    sum2 = sum([prefs[p2][it] for it in si])

    # Sums of the squares
    sum1Sq = sum([pow(prefs[p1][it], 2) for it in si])
    sum2Sq = sum([pow(prefs[p2][it], 2) for it in si])

    # Sum of the products
    pSum = sum([prefs[p1][it] * prefs[p2][it] for it in si])

    # Calculate r (Pearson score)
    num = pSum - sum1 * sum2 / n
    den = sqrt((sum1Sq - pow(sum1, 2) / n) * (sum2Sq - pow(sum2, 2) / n))
    if den == 0:
        return 0
    r = num / den
    return r

In [ ]:
def topMatches(
    prefs,
    person,
    n=5,
    similarity=sim_pearson,
):
    '''
    Returns the best matches for person from the prefs dictionary.
    Number of results and similarity function are optional params.
    '''

    scores = [(similarity(prefs, person, other), other) for other in prefs
              if other != person]
    scores.sort()
    scores.reverse()
    return scores[0:n]

In [ ]:
topMatches(critics, 496, n = 943)

 These are the users that are the least correlated

 (-1.0, 206),
 (-1.0, 205),
 (-1.0, 126),
 (-1.0, 39),
 (-1.0, 19),
 (-1.000000000000004, 672)

In [ ]:
#to find ALL the ratings for all the film that I have not seen:


In [ ]:
critics

In [29]:
def transformPrefs(prefs):
    '''
    Transform the recommendations into a mapping where persons are described
    with interest scores for a given title e.g. {title: person} instead of
    {person: title}.
    '''

    result = {}
    for person in prefs:
        for item in prefs[person]:
            result.setdefault(item, {})
            # Flip item and person
            result[item][person] = prefs[person][item]
    return result

In [30]:
movies = transformPrefs(critics)

In [59]:
topMatches(movies, 921, n = 1682)[-5:]

[(-1.0, 453), (-1.0, 452), (-1.0, 388), (-1.0, 376), (-1.000000000000004, 844)]